### PR TITLE
Support SQL standard "delete from ... where ..." syntax and lightweight implementation on merge tree tables

### PR DESCRIPTION
**Pull request by zhangjmruc**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Support SQL standard DELETE FROM syntax on merge tree tables and lightweight delete implementation for merge tree families.

Description:
This is planned for lightweight delete on merge tree families. In this pull request, we can parse SQL standard DELETE syntax, and the lightweight delete is converted to a normal update like "update _row_exists = 0 where predicate". 
    There will be an _row_exists system virtual column stored in data part. 
    The SELECT query and MERGE now apply the deleted rows mask implicitly, considered as PREWHERE predicate, filtering the "_row_exists = 0" rows.

=== Tests for lightweight delete and normal delete with "mutations_sync = 1"
The table has 12 columns with different data types, total rows are 100,000,000, about 110 parts.
The results show that the lightweight delete runs faster than normal delete. 200ms vs 8s.

--- 15 single row deletes
企业微信截图_39322126-b3bc-495e-9ea9-6db1da27418e

--- 10 range rows deletes
企业微信截图_d1dd0c0c-1dd5-48c0-a4e9-638c11a245f2

### CODE REVIEW

**Review by alesapin — CHANGES_REQUESTED** [review]
From my point of view, such syntax will be more confusing for users. `DELETE`s in clickhouse is extremely slow and expensive operations. When the user sees `ALTER` prefix in `ALTER DELETE` query it became more clear that it's not an ordinary OLTP database delete operation and maybe you have to execute them in a special way. So I think we shouldn't introduce this new standard syntax until we make our `DELETE`s faster.

**Comment by zhangjmruc** [comment]
@alexey-milovidov, I tried to implemented a prototype for lightweight delete + sql standard syntax. 
The deleted mask is now stored as String. 
For select query, the deleted mask filter works like row filter, the deleted rows are filtered implicitly. 
Please help to add "can be tested"

**Comment by zhangjmruc** [comment]
@alesapin I implemented a version of lightweight delete. You left a change requested, however I cannot resolve it. So the merging is blocked. Please help to add a "can be tested" tag, so I can continue the work.

**Comment by lqhl** [comment]
When we were using ClickHouse, we were not satisfied with the performance of deletion, so we developed the lightweight delete feature using lazy deletion. After preliminary testing, we found that the deletion performance has been greatly improved compared to the original `alter table ... delete`. We want to contribute our lightweight delete solution to the community to help ClickHouse meet more user needs. Can you mark this part of the code as `can be tested`, so that we can conduct a comprehensive test? If there are functional or performance problems, we are willing to continue to revise according to the opinions of the community. We also plan to continue to submit features such as lightweight delete on ReplicatedMergeTree on the basis of this feature.

@alesapin

**Comment by zhangjmruc** [comment]
@alesapin The failed tests are not related to this feature. Would you please take a look at the code changes for lightweight delete on merge tree with only wide part? Let me know if you have any comments.

**Comment by zhangjmruc** [comment]
Summary of 3 failing checks:
1. Stress test (memory, actions): table test_15.tp_2 is a replicated merge tree. lightweight delete doesn't kick in.
2022.07.04 07:49:09.157043 [ 542812 ] {} <Error> test_15.tp_2 (debc2ecd-e9b2-497c-8c48-7ec57d5d4258): Cannot quickly remove directory /var/lib/clickhouse/store/deb/debc2ecd-e9b2-497c-8c48-7ec57d5d4258/delete_tmp_all_0_0_0_2 by removing files; fallback to recursive removal. Reason: Code: 458. DB::ErrnoException: Cannot unlink file /var/lib/clickhouse/store/deb/debc2ecd-e9b2-497c-8c48-7ec57d5d4258/delete_tmp_all_0_0_0_2/pp.proj, errno: 21, strerror: Is a directory. (CANNOT_UNLINK) (version 22.7.1.1)
...
Referenced from error log:
2022.07.04 07:31:59.987510 [ 232572 ] {} <Error> test_15.tp_2 (debc2ecd-e9b2-497c-8c48-7ec57d5d4258): void DB::StorageReplicatedMergeTree::mutationsFinalizingTask(): Code: 242. DB::Exception: Table is in readonly mode (replica path: /clickhouse/tables/s1/01710_projection_fetch_test_15/replicas/2_r1). (TABLE_IS_READ_ONLY)

2. Stress Test (address, Actions): clieckhouse-server failed to start due to former server not exit.
Referenced from runlog.log:
+ echo 'Cannot start clickhouse-server'
+ echo -e 'Cannot start clickhouse-server\tFAIL'
+ cat /var/log/clickhouse-server/stdout.log
Cannot start clickhouse-server
/var/run/clickhouse-server/clickhouse-server.pid file exists and contains pid = 460221.
The process with pid = 460221 is already running.

3. Stateless tests (thread, actions) [1/3] -- not related, the failed test contains only insert and select. 

00731_long_merge_tree_select_opened_files | FAIL | 97.64
-- | -- | --
2022-07-04 00:40:48 Received exception from server (version 22.7.1): 2022-07-04 00:40:48 Code: 81. DB::Exception: Received from localhost:9000. DB::Exception: Database test_2vos10 doesn't exist. (UNKNOWN_DATABASE) 2022-07-04 00:40:48 (query: SYSTEM FLUSH LOGS)

File: src/Interpreters/InterpreterDeleteQuery.cpp
Comment (alesapin) [inline src/Interpreters/InterpreterDeleteQuery.cpp:64]: ?????

File: src/Storages/MergeTree/MutateTask.cpp
Comment (alesapin) [inline src/Storages/MergeTree/MutateTask.cpp:1736]: I thought that `files_to_skip` means "files to hard link, not to mutate", but ok, maybe I'm wrong.

File: src/Storages/MergeTree/MutateTask.cpp
Comment (alesapin) [inline src/Storages/MergeTree/MutateTask.cpp:1358]: Some comment, why we really need additional task here

File: src/Storages/MergeTree/MutateTask.cpp
Comment (alesapin) [inline src/Storages/MergeTree/MutateTask.cpp:1364]: why throw exceptions in these methods?

File: src/Interpreters/InterpreterDeleteQuery.cpp
Comment (alesapin) [inline src/Interpreters/InterpreterDeleteQuery.cpp:33]: Let's introduce a setting like `allow_experimental_lightweight_delete` which is disabled by default.

File: src/Interpreters/InterpreterDeleteQuery.cpp
Comment (alesapin) [inline src/Interpreters/InterpreterDeleteQuery.cpp:79]: Looks like this condition is useless.

File: src/Interpreters/MutationsInterpreter.h
Comment (alesapin) [inline src/Interpreters/MutationsInterpreter.h:47]: I don't like default arguments like this.

File: src/Parsers/ParserDeleteQuery.h
Comment (alesapin) [inline src/Parsers/ParserDeleteQuery.h:9]: I'd like to have setting `lightweight_delete_mutation = 0/1` to make ordinary `ALTER DELETE` queries lightweight. And enable it by default to check all tests.

File: src/Storages/MergeTree/IMergeTreeDataPart.cpp
Comment (alesapin) [inline src/Storages/MergeTree/IMergeTreeDataPart.cpp:1214]: Why?? We should be able to mutate compact parts in lightweight mode.

File: src/Storages/MergeTree/MergeTreeDataPartDeletedMask.cpp
Comment (alesapin) [inline src/Storages/MergeTree/MergeTreeDataPartDeletedMask.cpp:27]: Strange name.

**Review by alesapin — COMMENTED** [review]
Great contribution, thank you! Please check my review remarks. If possible, please add comments directly to the code instead of answering in PR.

I'll ask @davenger to check read part.

File: src/Parsers/ASTDeleteQuery.cpp
Comment (zhangjmruc) [reply src/Parsers/ASTDeleteQuery.cpp:9]: When I tried to remove the () of delim + getDatabase(), the following compile error came up:
/home/ubuntu/opensource/fork/ClickHouse/src/Parsers/ASTDeleteQuery.cpp:9:26: error: adding 'char' to a string does not append to the string [-Werror,-Wstring-plus-int]
    return "DeleteQuery" + delim + getDatabase() + delim + getTable();
           ~~~~~~~~~~~~~~^~~~~~~

File: src/Storages/IStorage.h
Comment (zhangjmruc) [reply src/Storages/IStorage.h:239]: I mean that the table contains at least one data part that has lightweight deleted mask file, data was already modified by LWD. The purpose here is to disable count() optimization.

File: src/Storages/MergeTree/MergeTreeData.cpp
Comment (zhangjmruc) [reply src/Storages/MergeTree/MergeTreeData.cpp:5669]: This is intended to fix that "select count() from table" return wrong rows count. Honestly, I didn't dig further why disable of count optimization doesn't work.

File: src/Storages/MergeTree/ReplicatedMergeTreeQueue.cpp
Comment (zhangjmruc) [reply src/Storages/MergeTree/ReplicatedMergeTreeQueue.cpp:1923]: Currently lightweight delete is not supported for ReplicatedMergeTree. Will be implemented in future.

File: src/Interpreters/InterpreterDeleteQuery.cpp
Comment (zhangjmruc) [reply src/Interpreters/InterpreterDeleteQuery.cpp:43]: Replicated tables will support lightweight delete in the next version. In this pr, focus on merge tree.

File: src/Storages/MergeTree/MutateTask.cpp
Comment (zhangjmruc) [reply src/Storages/MergeTree/MutateTask.cpp:1736]: files_to_skip means files not to hard link, like, checksums.txt, columns.txt ... 
Referenced from MutateSomePartColumnsTask's prepare():
...
        NameSet hardlinked_files;
        /// Create hardlinks for unchanged files
        for (auto it = ctx->source_part->data_part_storage->iterate(); it->isValid(); it->next())
        {
            if (ctx->files_to_skip.contains(it->name()))
                continue;
....

File: src/Storages/MergeTree/MergeTreeRangeReader.cpp
Comment (zhangjmruc) [reply src/Storages/MergeTree/MergeTreeRangeReader.cpp:997]: We store ColumnUInt8 plus some headers on disk now, the codes for deleted mask file are referenced from #32774. In the deleted mask, 1 marks for deleted rows, however in filter, 0 marks for filtered rows. Hence we need to some translation for read to filter out the deleted rows.

File: src/Storages/MergeTree/MergeTreeRangeReader.cpp
Comment (zhangjmruc) [reply src/Storages/MergeTree/MergeTreeRangeReader.cpp:1006]: The logic here is exactly same as _part_offset (logical row number in part). For deleted mask we need to get deleted rows with the same offset of read rows.

File: src/Storages/MergeTree/MutateTask.cpp
Comment (zhangjmruc) [reply src/Storages/MergeTree/MutateTask.cpp:1364]: Referenced from MutateSomePartColumnsTask, not sure why.

File: src/Storages/MergeTree/MergeTreeDataPartDeletedMask.h
Comment (zhangjmruc) [reply src/Storages/MergeTree/MergeTreeDataPartDeletedMask.h:20]: "const static" caused compile error. Hence use const.

**Comment by Algunenano** [comment]
Question about the feature in general:

> A new data part is created after lightweight delete, all files expect checksums, deleted_rows_mask.bin... are created from hard links of the source part. The deleted rows mask is updated based on the old one in source part + new deleted rows's _part_offset according to where conditions.

How does this work with projections? Hard linking the data from the projections will lead to incorrect data. Shouldn't they be regenerated every time? Or maybe invalidated? Or maybe for now throw an error as not supported.

**Comment by davenger** [comment]
@zhangjmruc Thank you very much for working on this feature! I looked through the implementation and I'd like to suggest some improvements.

As far as I see from the code the whole deleted_rows mask for a part is always loaded in memory with the part metadata. 
This might consume significant amount of memory for big tables.

The other thing is about filtering. Making the deleted_rows be more like an ordinary column with corresponding .mrk file should benefit it. 
Lets say we have a table like this

```
CREATE TABLE t (id UInt64 , value String) ENGINE MergeTree() ORDER BY id;

INSERT INTO t SELECT number, randomString(1000) FROM system.numbers LIMIT 1000000;
```

where values are rather big strings
And at some point we delete a range of rows that covers several granules

```
DELETE FROM t WHERE id < 100000
```

and then we read form the table

```
SELECT id, length(value) FROM t ORDER BY id LIMIT 3

Query id: b70bf4f9-e882-47eb-9ba6-cc82a8a3cdca

┌─────id─┬─length(value)─┐
│ 100000 │          1000 │
│ 100001 │          1000 │
│ 100002 │          1000 │
└────────┴───────────────┘

3 rows in set. Elapsed: 0.561 sec. Processed 106.50 thousand rows, 6.61 MB (189.89 thousand rows/s., 11.78 MB/s.)
```

This is rather slow because we don't filter whole granules where deleted_mask is 1 but read big `value` column  

if we add a fake "always true" PREWHERE condition the query executes much faster: 

```
SELECT id, length(value) FROM t PREWHERE id >= 0 ORDER BY id LIMIT 3

Query id: 989e1721-31f2-4f8b-84e2-e08a6ad581ba

┌─────id─┬─length(value)─┐
│ 100000 │          1000 │
│ 100001 │          1000 │
│ 100002 │          1000 │
└────────┴───────────────┘

3 rows in set. Elapsed: 0.087 sec. Processed 139.26 thousand rows, 42.44 MB (1.59 million rows/s., 486.05 MB/s.)
```

Here it first reads the thin `id` column, filters it by the mask and only then reads big `value` columns only for needed rows   

In order to have this fast filtering the deletion_mask check should be the first step in PREWHERE chain. I recently refactored reader code to support multiple steps of PREWHERE (#37165) and I'm willing to help with implementing this deletion_maks filtering.

**Comment by zhangjmruc** [comment]
> Great contribution, thank you! Please check my review remarks. If possible, please add comments directly to the code instead of answering in PR.
> 
> I'll ask @davenger to check read part.

@alesapin Please help to review parts of code changes and check if they are expected or not. 
The failed test case in integration tests are expected when we make ALTER DELETE lightweight, because the delete and drop column are mutated separately, hence the count of hard link are different as normal.
For the MergeTreeDataPartDeletedMask, to be honest, I'm not familiar, hence I only change the strange name of HashValue, and keep the char type instead of UInt128. From the review feedback of davenger, there might be some changes on this class.

**Comment by zhangjmruc** [comment]
> Question about the feature in general:
> 
> > A new data part is created after lightweight delete, all files expect checksums, deleted_rows_mask.bin... are created from hard links of the source part. The deleted rows mask is updated based on the old one in source part + new deleted rows's _part_offset according to where conditions.
> 
> How does this work with projections? Hard linking the data from the projections will lead to incorrect data. Shouldn't they be regenerated every time? Or maybe invalidated? Or maybe for now throw an error as not supported.

I disabled lightweight delete mutate for part with projections. Thank you for pointing this.

**Comment by zhangjmruc** [comment]
@davenger Thank you for suggesting improvements. Indeed, loading the deleted_rows mask with metadata will consume memory for big table. Hope if we plan to make deleted_rows like an ordinary column, this can be avoided.
Please help me to implement the delete_rows filter.

My concerns here are: 
Let's say the deleted_rows mask is stored in _delete.bin. 
1. Is the column _delete a virtual column? 
2. When should we add the "_delete = 0" to  PREWHERE? For all the selects on MergeTreeTables?   
3. How we read the _delete column in IMergeTreeReader? If we adding this to PREWHERE's read column, this is not a physical column in columns.txt.

>
> The other thing is about filtering. Making the deleted_rows be more like an ordinary column with corresponding .mrk file should benefit it. 

> In order to have this fast filtering the deletion_mask check should be the first step in PREWHERE chain. I recently refactored reader code to support multiple steps of PREWHERE (#37165) and I'm willing to help with implementing this deletion_maks filtering.

File: src/Storages/MergeTree/MergeTreeDataPartDeletedMask.cpp
Comment (zhangjmruc) [reply src/Storages/MergeTree/MergeTreeDataPartDeletedMask.cpp:30]: Keep char data type, as referenced from #32774.

**Comment by zhangjmruc** [comment]
@alesapin Following are the summary of the failing integration tests:
1. test_mutations_hardlinks/test.py::test_delete_and_drop_mutation : When we make ALTER DELETE lightweight by default, the delete and drop column are mutated separately, hence the count of hard link are different as normal. [We will remove the settings lightweight_delete_mutation, then this failing can be ignored.]
2. test_storage_meilisearch : New engine for MeiliSearch, I think they are not related to lightweight delete. 
3. test_log_family_hdfs : The error is due to "Exception: Can't wait HDFS to start". Not related to lightweight delete.

Please help to review the code changes as you suggested and let me know your opinions. In this PR we prefer to use UInt8   native format with proper headers to store deleted_rows mask which is loaded to memory when needed, and consider this as a prototype implementation for lightweight delete. Further improvements, like store as normal column for deleted_rows mask, support compact part ... can be implemented in the next version.

**Comment by lqhl** [comment]
Hi @alesapin @davenger , we haven't heard from you for a few days, can you tell us what is blocking this PR now?
We want to merge this PR first and then implement improvements and advanced features based on it.

**Comment by lqhl** [comment]
> Hi @alesapin @davenger , we haven't heard from you for a few days, can you tell us what is blocking this PR now? We want to merge this PR first and then implement improvements and advanced features based on it.

@alexey-milovidov

**Comment by zhangjmruc** [comment]
@alesapin As you suggested, I added a setting allow_experimental_lightweight_delete to control lightweight delete, the default value is false (Disabled). I also added another setting lightweight_delete_mutation to make ordinary ALTER DELETE queries lightweight to fully test lightweight delete. Please let me know if you have further review opinions?
    As for the format of deleted_rows mask, it can be stored in Native UInt8 format in this basic version. We can improve it with the help of davenger based on this version. 
    Please share me with any process or concerns on this. Thanks in advance.

**Comment by davenger** [comment]
Hi @zhangjmruc @lqhl! I've reworked the deleted mask logic. It took me a bit longer than I expected so I apologize for the slow response.
Here is the summary of the changes:
 - A new virtual column _row_exists serves as a deleted mask. Its value is inverted to simplify the filtering condition. When some rows are deleted this column is persisted as a regular column with the corresponding marks so it doesn't require special serialization logic.
 - "DELETE WHERE predicate" query is translated into "UPDATE _row_exists=0 WHERE predicate" 
 - SELECT queries on the parts with non-empty deletion masks get an extra filtering step equivalent to "PREWHERE _row_exists". This approach allows to skip reading granules where all rows are deleted.

**Comment by davenger** [comment]
> From my point of view, such syntax will be more confusing for users. `DELETE`s in clickhouse is extremely slow and expensive operations. When the user sees `ALTER` prefix in `ALTER DELETE` query it became more clear that it's not an ordinary OLTP database delete operation and maybe you have to execute them in a special way. So I think we shouldn't introduce this new standard syntax until we make our `DELETE`s faster.

This was address by introducing allow_experimental_lightweight_delete setting that is turned off by default

**Review by davenger — APPROVED** [review]

**Comment by zhangjmruc** [comment]
Hi @davenger, thank you for helping reworked lightweight delete logic! The deleted mask is stored as normal columns and SELECT queries can work faster with PREWHERE condition on deleted mask. 
However, as for the performance, because the DELETE WHERE queries are translated to UPDATE _row_exists = 0 WHERE, and the lightweight delete task is removed, now the lightweight delete is treated as an ordinary update mutation on the system column _row_exists. For cases that we delete 1 row in part with large number of rows, how is the performance of lightweight delete execution? Also for mutate all columns cases, the deleted mask is not applied, because PREWHERE on deleted mask can make faster?
For lightweight delete, the system table mutations and mutation.txt show the its command as "UPDATE _row_exists = 0 WHERE predicate". Can this be used as mark of lightweight delete?

**Review by alesapin — APPROVED** [review]

File: src/Storages/MergeTree/MutateTask.cpp
Comment (alesapin) [inline src/Storages/MergeTree/MutateTask.cpp:1487]: Maybe elaborate on this line. It's unclear how we will preserve previous delete masks if they existed. So let's clarify the comment.

**Review by alesapin — APPROVED** [review]

**Comment by alesapin** [comment]
We still need to implement replicated case

File: src/Storages/MergeTree/MutateTask.cpp
Comment (davenger) [reply src/Storages/MergeTree/MutateTask.cpp:1488]: Wrote a more detailed comment

**Comment by lqhl** [comment]
> We still need to implement replicated case

@alesapin We already implemented lightweight delete for ReplicatedMergeTree. Shall we submit the code to this PR or we can merge this one first and start a new PR for the replicated case.

**Comment by zhangjmruc** [comment]
@davenger @alesapin I tried to implement lightweight delete for replicated case and added a test case for it. Please help to review and let me know your opinion?

**Review by alesapin — APPROVED** [review]
I like the simplicity of implementation for RMT. Thanks!

**Comment by davenger** [comment]
2 perf test failures are expected because this new setting is not present in old build:
`lightweight_delete | DB::Exception: Unknown setting allow_experimental_lightweight_delete.
`

**Comment by WillMeng** [comment]
> ### Changelog category (leave one):
> * Improvement
> 
> ### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
> Support SQL standard DELETE FROM syntax on merge tree tables and lightweight delete implementation for wide part only.
> 
> Description: This is planned for lightweight delete on merge tree families. In this pull request, we can parse SQL standard DELETE syntax, and the lightweight delete is implemented for wide part only. A new data part is created after lightweight delete, all files expect checksums, deleted_rows_mask.bin... are created from hard links of the source part. The deleted rows mask is updated based on the old one in source part + new deleted rows's _part_offset according to where conditions. The SELECT query and MERGE now apply the deleted rows mask implicitly, filtering the "_delete = 1" rows.
> 
> === Tests for lightweight delete and normal delete with "mutations_sync = 1" The table has 12 columns with different data types, total rows are 100,000,000, about 110 parts. The results show that the lightweight delete runs faster than normal delete. 200ms vs 8s.
> 
> --- 15 single row deletes 企业微信截图_39322126-b3bc-495e-9ea9-6db1da27418e
> 
> --- 10 range rows deletes 企业微信截图_d1dd0c0c-1dd5-48c0-a4e9-638c11a245f2

Thanks for the feature. We used a lot of "SETTINGS min_bytes_for_wide_part = '100M'" in our table, when support Compact part?

**Comment by zhangjmruc** [comment]
@WillMeng I will update descriptions. The compact part and Replicated MergeTree are all supported. Please reference the new test cases.

**Comment by WillMeng** [comment]
> @WillMeng I will update descriptions. The compact part and Replicated MergeTree are all supported. Please reference the new test cases.

Thank you. I have another question, when the data by marked "deleted" is clear from disk.

**Comment by zhangjmruc** [comment]
> > @WillMeng I will update descriptions. The compact part and Replicated MergeTree are all supported. Please reference the new test cases.
> 
> Thank you. I have another question, when the data by marked "deleted" is clear from disk.

@WillMeng The "deleted" rows will be cleared after background merge.

**Comment by WillMeng** [comment]
@zhangjmruc  , Hi, We had try out this feature at first time. We found those problem:
1. SELECT COUNT() FROM  `system`.mutations WHERE is_done = 0, The result is 92568, and read timed out(30 seconds) at often
2. The destination  table become so slow when read it.

**Comment by zhangjmruc** [comment]
@WillMeng Would you please share us more details about the unfinished mutations? The column latest_fail_reason contains info about failure reason. Or you need to provide error messages in /var/log/clickhouse-server/clickhouse-server.err.log. 
Please open a new issue for your problem, we can continue inv there.

The lightweight delete just converted the delete to an update on virtual column _rows_exists. Not sure why so many unfinished mutations.

**Pull request merged**
